File: apps/backend/app/models/message.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import CheckConstraint, ForeignKey, String, Text, text
from sqlalchemy.dialects.postgresql import TIMESTAMP, UUID
from sqlalchemy.orm import Mapped, mapped_column

from ..db import Base


class Message(Base):
    __tablename__ = "messages"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, server_default=text("gen_random_uuid()")
    )
    trip_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("trips.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    body: Mapped[str] = mapped_column(Text, nullable=False)
    kind: Mapped[str] = mapped_column(String(20), nullable=False, default="text")
    sent_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=text("now()"), nullable=False
    )

    __table_args__ = (
        CheckConstraint(
            "kind in ('text','system','arrival','leave','join')",
            name="messages_kind_check",
        ),
    )

File: apps/mobile/lib/shared/models/message.dart
class MessageDto {
  const MessageDto({
    required this.id,
    required this.tripId,
    required this.userId,
    required this.body,
    required this.kind,
    required this.sentAt,
  });

  final String id;
  final String tripId;
  final String userId;
  final String body;
  final String kind;
  final DateTime sentAt;

  bool get isSystem => kind != 'text';

  factory MessageDto.fromJson(Map<String, dynamic> json) => MessageDto(
        id: json['id'] as String,
        tripId: json['trip_id'] as String,
        userId: json['user_id'] as String,
        body: json['body'] as String,
        kind: json['kind'] as String,
        sentAt: DateTime.parse(json['sent_at'] as String),
      );
}
